### PR TITLE
[debugging feature] Allow raw loader selection by file extension

### PR DESCRIPTION
**Pull request by victoryforce**
Sometimes it would be nice to be able to temporarily select a different raw loader (libraw, that is) for debugging (to check the correctness of the white point, whether problem in loading the file are the result of corrupted file or a bug in rawspeed, etc.).

I believe that this PR is completely safe because of the simplicity of the code, the simplicity of the idea behind it, and the fact that when the user doesn't try to change the loader, everything works _exactly the same as before_. This code simply adds a feature that advanced users (and even dt devs) _could use_ if needed.

As an afterword: when testing this PR, I alternately loaded a considerable number of formats with both raw loaders, in some cases the higher speed of the rawspeed loader was visually noticeable!

And finally, I would like to emphasize once again that this PR should not be considered as an attempt to add the ability to choose a loader for all users during normal use of dt. __This is only a debugging aid__. That is why the interface for managing this feature was __intentionally__ not added to the dt settings.

### CODE REVIEW

**Comment by TurboGit** [comment]
I'm all for such debugging option, but I'm wondering if the solution here is just to be able to order libraw before rawspeed in imageio. I'm not sure we want to deal with extension at all, it will be simpler code wise and for debugging it will do the same.

**Comment by kmilos** [comment]
> be able to order libraw before rawspeed in imageio

+1

**Comment by kmilos** [comment]
One thing to note: LibRaw might (and probably will) return a slightly different crop, which might have some unexpected/unknown effects...

**Comment by victoryforce** [comment]
> I'm all for such debugging option, but I'm wondering if the solution here is just to be able to order libraw before rawspeed in imageio. I'm not sure we want to deal with extension at all, it will be simpler code wise and for debugging it will do the same.

But this code has already been written and already tested, it is simple enough (I would even say trivial) and for me it is no more complicated than just swapping loaders in imageio. _From my point of view, more flexibility with the same code complexity is always better_.

In addition (and this is also an important use case), for users who know what they are doing, it can be a ___temporary___ solution in a situation where they do not want to give up darktable, but the support of their new and shiny camera in rawspeed lags behind libraw. Reordering the loaders is much worse in this case, if only because for some other cameras we may have huge problems with libraw (for example, Nikon Df images do not open in the darkroom after import).

In summary, I see no advantage in abandoning this code and replacing it with an inflexible changing of loaders order.

**Comment by victoryforce** [comment]
> One thing to note: LibRaw might (and probably will) return a slightly different crop, which might have some unexpected/unknown effects...

Yes, it's a known fact. I can only emphasize once again that the purpose of this PR is NOT to replace one loader with another, but to create a flexible mechanism for those to whom it is suitable and who know why they need it.

**Comment by jenshannoschwalm** [comment]
I would prefer another debugging option '-d libraw' to handle this alternative loader. It should be handled as the verbose option so not included by '-d all'.

**Comment by victoryforce** [comment]
@TurboGit  I've tried to clearly describe the problem my PR solves and the use cases that call for that solution (rather than an unnecessarily oversimplified one that just reorders the loaders). I am afraid that I will hardly be able to add any arguments for this PR to be accepted. I can only emphasize once again that this simple and safe fix solves a real problem that many users are suffering from. It will help those users who currently have to change the extension of their raw files to CR3 only to be processed by libraw (a famous "rename to .CR3" hack advised on reddit and discuss.pixls.us).

I continue to believe that inclusion of this improvement in darktable 4.2.1 will be justified, safe and useful.

**Comment by jenshannoschwalm** [comment]
> I'm all for such debugging option, but I'm wondering if the solution here is just to be able to order libraw before rawspeed in imageio. I'm not sure we want to deal with extension at all, it will be simpler code wise and for debugging it will do the same.

I tested that and indeed a very simple solution. (using a global --enforce-libraw setting). What i observed: the rawprepare module does **not** get correct cropping, white- and blacklevel settings from libraw!

So not as useful for debugging as expected from my side. (white level seems to be very interesting ...)

File: src/imageio/imageio_libraw.c [hunk 1]
```diff
@@ -208,15 +208,31 @@ const model_map_t modelMap[] = {
...
   if(!ext) return FALSE;
   ext++;
-  for(const char **i = extensions_whitelist; *i != NULL; i++)
-    if(!g_ascii_strncasecmp(ext, *i, strlen(*i)))
-    {
-      return TRUE;
-    }
+
+  if(dt_conf_key_not_empty("libraw_extensions"))
+    extensions_whitelist = g_strjoin(" ", always_by_libraw, dt_conf_get_string_const("libraw_extensions"), NULL);
```
Comment (TurboGit) [inline src/imageio/imageio_libraw.c:221]: Let at least add an entry into darktableconfig.xml.in about this new config. The description will be helpful and document the purpose.

**Comment by kmilos** [comment]
> What i observed: the rawprepare module does **not** get correct cropping, white- and blacklevel settings from libraw!
> 
> So not as useful for debugging as expected

Yep, libraw can put black and white levels (and color matrix and crop) in different fields of its exposed data structure depending on the camera vendor (and possibly model/generation). What we wired up and tested so far is specific to Canon (CR3s)... So it's not a 1:1 general replacement in the current state of things unfortunately.

File: src/imageio/imageio_libraw.c (hunk 1 above)
Comment (victoryforce) [reply src/imageio/imageio_libraw.c:221]: OK, done.

**Review by TurboGit — APPROVED** [review]
Thanks!
